cpp/oneapi/dal/backend/primitives/selection/kselect_by_rows_heap_dpc.cpp

```diff
@@ -0,0 +1,448 @@
+        idx_t pbuff_ids[pbuff_size] = { idx_default };
+        std::int32_t pbuff_count, prev_count;
+
+        sel_t* const heaps = heaps_.get_pointer();
+        sel_t* const curr_heap = heaps + (k_ + 1) * sid;
+        sel_t* const curr_heap_end = curr_heap + (k_ + 1);
+
+        // Heap initialization
+        for (std::int32_t i = cid; i <= k_; i += sg_width) {
+            auto& values = curr_heap[i];
+            values.idx = idx_default;
+            values.dst = dst_default;
+        }
+        sg.barrier();
+
+        std::int32_t k_written = 0;
+        dst_t worst_val = dst_default;
+        const auto step = sg_width * pbuff_size;
+        const auto* const row = src_ptr_ + rid * src_str_;
+        for (std::int32_t i = 0; i < width_; i += step) {
+            const std::int32_t block_start_col = cid + i;
+
+            pbuff_count = 0, prev_count = 0;
+            worst_val = curr_heap->dst;
+
+            // Collecting temporary best values in private memory
+            for (std::int32_t j = 0; (j < pbuff_size); ++j) {
+                const idx_t idx = block_start_col + sg_width * j;
+                const bool handle = idx < width_;
+                const dst_t val = handle ? *(row + idx) : dst_default;
+                pbuff_count += bool(val < worst_val);
+                pbuff_ids[prev_count] = idx;
+                pbuff_dst[prev_count] = val;
+                prev_count = pbuff_count;
+            }
+            sg.barrier();
+
+            // writing temporary best values into heap
+            std::int32_t cid_to_handle, handle_this;
+            do {
+                handle_this = pbuff_count ? cid : -1;
+                k_written = reduce(sg, k_written, max_func);
+                cid_to_handle = reduce(sg, handle_this, max_func);
+                if (cid_to_handle == cid) {
+                    for (std::int32_t i = 0; i < pbuff_count; ++i, ++k_written) {
+                        worst_val = curr_heap->dst;
+                        const auto curr_val = pbuff_dst[i];
+                        const bool handle = curr_val < worst_val;
+                        sel_t result{ std::move(curr_val), pbuff_ids[i] };
+                        if (handle) {
+                            if (k_ > k_written) {
+                                *(curr_heap + k_written) = std::move(result);
+                                push_heap(curr_heap, curr_heap + k_written);
+                            }
+                            else {
+                                replace_first(std::move(result), curr_heap, curr_heap + k_);
+                            }
+                        }
+                        k_written += handle;
+                    }
+                    pbuff_count = 0;
+                }
+                sg.barrier();
+            } while (cid_to_handle > -1);
+        }
+
+        // Sorting heap before writing out
+        // TODO: Think out if it can be performed in parallel
+        if (cid == 0) {
```

Comment:
K-Select doesn't need results to be sorted

Reply:
Is it? I mean it's not hard (in terms of complexity) to sort heap in such approach and it can be beneficial for testing

Reply:
Is it being sorted again for voting? Other selections don't provide sorted data